### PR TITLE
extract.gnm clean after several merged PR

### PR DESCRIPTION
**Pull request by czucca**
Hi @leifeld 

I copied here the original @string-token contribution and added the setmethod which I couldn't find in the original one. 

I'll check more carefully the formatting asap!

### CODE REVIEW

**Comment by czucca** [comment]
Hi Philip and string-token, 

I'm reviewing this extract function and wondering on the opportunity of including so many GOFs:

1) There was an error in calling Nobs. when correct nobs == ncell (which is one of the new arguments). Hence, I would not include ncell since it is Nobs. 

2) model p value. Many other models provide a p value for the model. It was never included before. Is it necessary to add? it is not even provided in model summary for class gnm.

3) sorry, my bad, what is this? 
AICL2?
BICL2?
google did not help me either to understand. 

4) Dissimilarity index:  not provided in model summary either. Even if this could be a useful thing to make sense of your results, I do not believe it can be inserted among the goodness of fit. This is a further stage of analysis in my opinion. 

Happy to include all them if @leifeld thinks they are useful. 

Hope this helps, 
Claudia

**Comment by leifeld** [comment]
Thanks, @czucca! Regarding your questions:
1. I don't really know what `ncell` is or does. Let's give @string-token a day or two to respond, and if we don't hear back, remove the `ncell` part.
2. Can you look at the original `summary` method in their package and see if/how they compute the p-values? It's sometimes the case that package authors decide not to store the results in the summary but just display them on the console when `summary` is called, meaning some things may be calculated ad-hoc. Perhaps you can find it there and include it. If not, let's leave them out.
3. I'm not very familiar with this either, but I think AIC can be interpreted as a regularization that penalizes larger numbers of model parameters, and I could imagine that L2 indicates L2 norm regularization? But that's just a shot in the dark... Perhaps @string-token has a better idea. In the absence of other information, we should call the respective arguments `include.aic` and `include.bic`.
4. I would have no problem with the inclusion of this statistic. The user can still switch it off if that's not needed...

Can you please also fix the formatting? Include spaces in the appropriate places etc. Also move the calculation of the GOF statistics into respective conditions further below such that they are not all executed even if the user switches them off... Thanks!

**Comment by leifeld** [comment]
No response... let's do it.

**Comment by czucca** [comment]
update: this class of models is a bit tricky. In most of the cases does not provide SE and Pvalues (just a list of NAs), and for some reasons that I'm still trying to understand, one of the internal functions of gnm is conflicting with texreg. So it will take me a while to fix this.

**Comment by czucca** [comment]
Hi Philip, 

I finally fund the problem and now the extract function works. I also introduced a correction. If I just print the output the tables comes out very shaggy:


```
======================================
                             Model 1  
--------------------------------------
(Intercept)                  4.20     
                                      
SES.L                        -0.06    
                                      
SES.Q                        -0.32    
                                      
SES.C                        -0.21    
                                      
SES^4                        0.18     
                                      
SES^5                        0.14     
                                      
MHS.L                        0.00     
                                      
MHS.Q                        0.00     
                                      
MHS.C                        0.39     
                             (0.07)   
Mult(., SES, MHS).           0.80     
                                      
Mult(1, ., MHS).SESA         0.00     
                                      
Mult(1, ., MHS).SESB         -0.02 ***
                             (0.15)   
Mult(1, ., MHS).SESC         0.18     
                             (0.15)   
Mult(1, ., MHS).SESD         0.23     
                             (0.14)   
Mult(1, ., MHS).SESE         0.59     
                             (0.22)   
Mult(1, ., MHS).SESF         1.00     
                                      
Mult(1, SES, .).MHSwell      -1.04    
                                      
Mult(1, SES, .).MHSmild      0.18     
                                      
Mult(1, SES, .).MHSmoderate  0.33     
                                      
Mult(1, SES, .).MHSimpaired  0.17     
                                      
--------------------------------------
df. residuals                10       
Log Likelihood               -92.42   
Deviance                     40.66    
Pearson chi-squared          37.95    
AIC                          212.83   
BIC                          229.33   
Dissim. Index                5.73     
Num. obs.                    24       
======================================
*** p < 0.001; ** p < 0.01; * p < 0.05
```
in fact, several variables do not have an error or a pvalue. The gnm packages provides its own inbuilt function to get rid of non relevant variables. Hence I repliecated the idea inside texreg that now is printing only variables with se and pvalue. 

```
===============================
                      Model 1  
-------------------------------
MHS.C                 0.39     
                      (0.07)   
Mult(1, ., MHS).SESB  -0.02 ***
                      (0.15)   
Mult(1, ., MHS).SESC  0.18     
                      (0.15)   
Mult(1, ., MHS).SESD  0.23     
                      (0.14)   
Mult(1, ., MHS).SESE  0.59     
                      (0.22)   
-------------------------------
df. residuals         10       
Log Likelihood        -92.42   
Deviance              40.66    
Pearson chi-squared   37.95    
AIC                   212.83   
BIC                   229.33   
Dissim. Index         5.73     
Num. obs.             24       
===============================
*** p < 0.001; ** p < 0.01; * p < 0.05
```

let me know if you are ok with this solution. 

I'll update the documentation asap, 

Claudia

**Comment by czucca** [comment]
note that I've set to false some GOFs in the final version, while all of them are printed above

File: R/extract.R [hunk 1]
```diff
@@ -5354,3 +5354,91 @@ extract.feglm <- function(model, include.deviance = TRUE, include.nobs = TRUE,
...
+    }
+    if (include.bic == TRUE) {
+        bic <- BIC(model)[1]
+        gof <- c(gof, bic)
+        gof.names <- c(gof.names, "BIC")
+        gof.decimal <- c(gof.decimal, TRUE)
+    }
+    if (include.delta == TRUE) {
+        delta <- sum(na.omit(c(abs(residuals(model, "response"))))) / 
+            sum(na.omit(c(abs(fitted(model)))))/2*100 # Dissimilarity index
```
Comment (leifeld) [inline R/extract.R]: Check the spacing here.

File: R/extract.R [hunk 2]
```diff
@@ -5354,3 +5354,91 @@ extract.feglm <- function(model, include.deviance = TRUE, include.nobs = TRUE,
...
+                        include.chisq = FALSE, include.delta = FALSE, ...) {
+    
+    s <- summary(model)
+    coefficients.names <- names(model$coefficients)
+    co <- s$coef[, 1]
+    se <- s$coef[, 2]
+    pval <- s$coef[, 3]
+
+    table <- as.data.frame(cbind(coefficients.names, co, se, pval))
+    table <- table[!is.na(table$se),]
```
Comment (leifeld) [inline R/extract.R]: Add a space after the comma,

**Comment by czucca** [comment]
all done! yeahh :)

File: R/extract.R [hunk 3]
```diff
@@ -5354,3 +5354,91 @@ extract.feglm <- function(model, include.deviance = TRUE, include.nobs = TRUE,
     return(tr)
 }
 setMethod("extract", signature = className("feglm", "alpaca"), definition = extract.feglm)
+                                         
+                                         # extension for gnm objects (gnm package)
```
Comment (leifeld) [inline R/extract.R]: What happened to the spacing here?

File: R/extract.R (hunk 1 above)
Comment (leifeld) [reply R/extract.R]: Please add spaces around `/` and `*`.

File: R/extract.R (hunk 3 above)
Comment (czucca) [reply R/extract.R]: I have no idea... I guess it's related to the conflicts...

**Comment by leifeld** [comment]
Thx. Feel free to merge.

**Comment by string-token** [comment]
Although I am a bit late and the merge has been done I would like to clarify the questions/issues by Claudia:

>     1. There was an error in calling Nobs. when correct nobs == ncell (which is one of the new arguments). Hence, I would not include ncell since it is Nobs.

The problem with standard R procedures to get the number of observations is that functions like `nobs(model)` or `length(model)` turn out the number of cells from the table instead of the actual total of observations which is what you normally would expect. Hence, the need to use `sum(model$y)`.

>     2. model p value. Many other models provide a p value for the model. It was never included before. Is it necessary to add? it is not even provided in model summary for class gnm.

You can often find p-values for RC models in the literature (see e.g. Wong 2010 https://books.google.de/books?id=0eNiR5IHrE0C&pg=PA113). Therefore, I included this "pmodel".

>     3. sorry, my bad, what is this?
>        AICL2?
>        BICL2?
>        google did not help me either to understand.

Again, the problem with `gnm` is that its focus on tables lies a bit square with R standard functions.  `AIC(model)` and `BIC(model)`, even though printed in the summary by gnm itself, will give you values which you will seldom find in the literature (e.g. Xie 1992). Normally, the AIC and BIC  for RC models are calculated based on the deviance of the model (=L²). The [logmult-package](https://github.com/nalimilan/logmult/blob/master/R/unidiff.R) provides some nice GOF statistics.

>     4. Dissimilarity index:  not provided in model summary either. Even if this could be a useful thing to make sense of your results, I do not believe it can be inserted among the goodness of fit. This is a further stage of analysis in my opinion.

This is frequently used in the literature (e.g. Wong 2010).


References:
Wong, R. S.-K. (2010). Association models. Los Angeles, CA: Sage.
Xie, Y. (1992). The Log-Multiplicative Layer Effect Model for Comparing Mobility Tables. American Sociological Review, 57, 380–395.

**Comment by leifeld** [comment]
@czucca: can you make any necessary changes? Thx!

**Comment by czucca** [comment]
Hi Philip, I'm finishing to move plotreg to ggplot. I'm almost done, just one last thing to solve and some checks left. As soon as I'll be done I'll fix this.

**Pull request merged**